lib/pos_blockchain.go

```diff
@@ -0,0 +1,1134 @@
+	}
+
+	// We run the full validation algorithm on the block.
+	return bc.validateAndIndexBlockPoS(block)
+}
+
+// isValidBlockPoS performs all basic validations on a block as it relates to
+// the Blockchain struct.
+func (bc *Blockchain) isValidBlockPoS(block *MsgDeSoBlock) error {
+	// Surface Level validation of the block
+	if err := bc.isProperlyFormedBlockPoS(block); err != nil {
+		return err
+	}
+	if err := bc.isBlockTimestampValidRelativeToParentPoS(block); err != nil {
+		return err
+	}
+	// Validate Block Height
+	if err := bc.hasValidBlockHeightPoS(block); err != nil {
+		return err
+	}
+	// Validate View
+	if err := bc.hasValidBlockViewPoS(block); err != nil {
+		// Check if err is for view > latest committed block view and <= latest uncommitted block.
+		// If so, we need to perform the rest of the validations and then add to our block index.
+		// TODO: implement check on error described above. Caller will handle this.
+		return err
+	}
+	return nil
+}
+
+// isBlockTimestampValidRelativeToParentPoS validates that the block's timestamp is
+// greater than its parent's timestamp.
+func (bc *Blockchain) isBlockTimestampValidRelativeToParentPoS(block *MsgDeSoBlock) error {
+	// Validate that the timestamp is not less than its parent.
+	parentBlock, exists := bc.blockIndexByHash[*block.Header.PrevBlockHash]
+	if !exists {
+		// Note: this should never happen as we only call this function after
+		// we've validated that all ancestors exist in the block index.
+		return RuleErrorMissingParentBlock
+	}
+	if block.Header.TstampNanoSecs < parentBlock.Header.TstampNanoSecs {
+		return RuleErrorPoSBlockTstampNanoSecsTooOld
+	}
+	return nil
+}
+
+// isProperlyFormedBlockPoS validates the block at a surface level. It checks
+// that the timestamp is valid, that the version of the header is valid,
+// and other general integrity checks (such as not malformed).
+func (bc *Blockchain) isProperlyFormedBlockPoS(block *MsgDeSoBlock) error {
+	// First make sure we have a non-nil header
+	if block.Header == nil {
+		return RuleErrorNilBlockHeader
+	}
+
+	// Make sure we have a prevBlockHash
+	if block.Header.PrevBlockHash == nil {
+		return RuleErrorNilPrevBlockHash
+	}
+
+	// Timestamp validation
+	// First make sure we have a non-nil header
+	if block.Header == nil {
+		return RuleErrorNilBlockHeader
+	}
+
+	// Make sure we have a prevBlockHash
+	if block.Header.PrevBlockHash == nil {
+		return RuleErrorNilPrevBlockHash
+	}
+
+	// Timestamp validation
+	// TODO: Add support for putting the drift into global params.
+	if block.Header.TstampNanoSecs > uint64(time.Now().UnixNano())+bc.params.DefaultBlockTimestampDriftNanoSecs {
+		return RuleErrorPoSBlockTstampNanoSecsInFuture
+	}
+
+	// Header validation
+	if block.Header.Version != HeaderVersion2 {
+		return RuleErrorInvalidPoSBlockHeaderVersion
+	}
+
+	// Malformed block checks
+	// All blocks must have at least one txn
+	if len(block.Txns) == 0 {
```

Comment:
I'm not sure if we need a block reward in PoS. I can't remember where we landed after discussing it, but right now validators are rewarded based on an APY and based on fees from txns. The APY is split between stakers and the validator they've delegated to based on the validator's commission. So what would we need a block reward for? Just as extra incentive for the leader I guess? Not crazy.

If we choose to keep the block reward let's make it adjustable via ParamUpdater, including setting it to zero if it's no longer needed in the future.

Reply:
@tholonious @AeonSw4n - we discussed it and were going to keep the block reward to reward the leader I believe.

Reply:
Ah sorry just saw this reply. Ok if we keep it can we make it based on a GlobalParam instead of on supply.go?